File: trunk/src/dotnet/yaxgl/yaxgl/EventHandlerManager.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Reflection;

namespace de.yaxgl
{
    public class EventHandlerManager
    {
        IList<EventHandlerMethod> eventHandlerMethods = null;

        public EventHandlerManager()
        {
            eventHandlerMethods = new List<EventHandlerMethod>();
        }
        

        public void registerEventHandlers(Object eventReciever)
        {
            MethodInfo[] methods = eventReciever.GetType().GetMethods();
            foreach (MethodInfo method in methods)
            {
                
                Object[] attributes = method.GetCustomAttributes(typeof(EventHandler), false);
                if (attributes.Length > 0)
                {
                    EventType eventType=((EventHandler)attributes[0]).eventType;
                    string eventID=((EventHandler)attributes[0]).eventID;
                    
                    // now we need to check whether the signature of the annotated method is long enough
				    if(method.GetParameters().Length != 2) {
					    //TODO: geeignete Exception werfen
                        throw new Exception("Event handler must have exactly 2 Arguments.");
				    }
    				
				    // we also need to check whether the signature of the annotated method is what we expect it to be
                    Type componentParameter = method.GetParameters()[0].ParameterType;
				    Type argParameter = method.GetParameters()[1].ParameterType;
				    if((componentParameter!=typeof(Component) || componentParameter != typeof(Component)) || argParameter != typeof(EventArgs))
                    {
                        throw new Exception("Event Handler must have 2 Arguments with Type " + typeof(EventType).ToString() + " and " + typeof(string).ToString() + ".");
                    }
                    
                    EventHandlerMethod eventHandlerMethod = new EventHandlerMethod(eventReciever, method, new EventArgs(eventType,eventID));
                    eventHandlerMethods.Add(eventHandlerMethod);
                }
            }
        }
        
        public void invokeHandlers(Component component, EventArgs eventArgs) {
		    foreach(EventHandlerMethod eventHandlerMethod in eventHandlerMethods) {
                if(eventHandlerMethod.isMatching(eventArgs)) {
				    eventHandlerMethod.invokeMethod(component,eventArgs);
			    }
		}
	}

    } 
} // END NAMESPACE

File: trunk/src/java/YaxglLibrary/de/yaxgl/EventDispatcher/EventHandlerManager.java
package de.yaxgl.EventDispatcher;

import java.lang.annotation.*;
import java.lang.reflect.*;
import java.util.*;

import de.yaxgl.Base.*;
import de.yaxgl.EventDispatcher.Attribute.*;

public class EventHandlerManager {
	List<EventHandlerMethod> eventHandlerMethods = null;
    
    public EventHandlerManager()
    {
        eventHandlerMethods = new ArrayList<EventHandlerMethod>();
    }
    
    public void registerEventHandlers(Object eventReciever) throws Exception
    {    
        for(Method method : eventReciever.getClass().getMethods())
        {
	    	Annotation[] annotations = method.getAnnotations();   
            if (annotations.length > 0)
            {
                for(Annotation annotation : annotations)
                {
                    if(annotation.getClass()==EventHandler.class)
                    {
                    	EventType eventType = ((EventHandler)annotation).eventType();
                    	String eventID = ((EventHandler)annotation).eventID();
                    	boolean regexOn = ((EventHandler)annotation).regexOn();
                    	
                    	// now we need to check whether the signature of the annotated method is long enough
                    	// 2 for nomal use 3 with option regexON=true or off=false standart is off
                    	if ((method.getParameterTypes().length < 2) && (method.getParameterTypes().length > 3))
                    	{
                    		//TODO: geeignete Exception werfen
                    		throw new Exception("Event handler must have exactly 2 Arguments.");
                    	}

                    // we also need to check whether the signature of the annotated method is what we expect it to be
                    Class componentParameter = method.getParameterTypes()[0];
                    Class argParameter = method.getParameterTypes()[1];
                    if (componentParameter != Component.class || argParameter != EventArgs.class)
                    {
                        throw new Exception("Event Handler must have 2 Arguments with Type " + EventType.class.toString() + " and " + String.class.toString() + ".");
                    }

                    EventHandlerMethod eventHandlerMethod = new EventHandlerMethod(eventReciever, method, eventType, eventID,regexOn);
                    eventHandlerMethods.add(eventHandlerMethod);
                    }
                 }
            }
        }
    }
    
    
    public void invokeHandlers(Component component, EventArgs eventArgs) {
	    for(EventHandlerMethod eventHandlerMethod : eventHandlerMethods) {
            if(eventHandlerMethod.isTypeOf(eventArgs.getType()) && eventHandlerMethod.isMatching(eventArgs.getID())) {
			    eventHandlerMethod.invokeMethod(component,eventArgs);
		    }
	}
}



}
